service/storage_proxy.cc

```diff
@@ -3760,8 +3760,8 @@ void storage_proxy::on_down(const gms::inet_address& endpoint) {
     auto it = _view_update_handlers_list->begin();
     while (it != _view_update_handlers_list->end()) {
         auto guard = it->shared_from_this();
-        if (it->get_targets().count(endpoint) > 0 && _response_handlers.find(it->id()) != _response_handlers.end()) {
-            it->timeout_cb();
+        if (it->get_targets().count(endpoint) > 0) {
+            got_failure_response(it->id(), endpoint, 1, {});
```

Comment:
@psarna I don't understand something here - timeout_cb() specifically in our case (CL=ANY) did hint_to_dead_endpoints() which is what we wanted to do - we wanted when the remote node goes down to quickly write hints. But I don't understand if got_failure_response() also writes hints. I couldn't find where it does, if it does, can you please help me find where?

Reply:
Sure. In case of CL=ANY, which we use for view updates, the flow for `get_failure_response` is:
1. Call `failure_response`, which returns true if CL is definitely not fulfilled and we should return an error to the client. This is never true for CL=ANY.
2. If `failure_response` returned false (always the case for CL=ANY), it calls `check_for_early_completion`
3. `check_for_early_completion` is just a named `if` statement: if there were as many failures as they are endpoints we wait for, call `timeout_cb()`.

The previous solution just called `timeout_cb()` unconditionally, which works fine enough without topology changes, because view updates are always sent to exactly 1 endpoint. With ongoing topology changes it may actually be sent to more, so instead of killing the whole handler as if all writes failed, only the failure for given endpoint is registered. So, `timeout_cb()` will only get triggered after all writes fail. In common case there's just one write, so `got_failure_response` will effectively simply call `timeout_cb()` as before, checking a few conditions first.

Reply:
Thanks. There's still some case I don't understand, but maybe it's a separate issue....
With your patch, in case we're also waiting for a second node, we will not call timeout_cb().
But a second later, when a response *does* come for that second node. Because we got one response, and wanted CL=ANY, this response will be enough. So that's it - we'll never write a hint for the dead node.
So I think it may be wrong to use CL=ANY here?
(again, this comment is not specifically about your patch, but about the current code in general)

Reply:
Yes, that's an interesting point, cl::ANY doesn't really seem to fit all cases well. I guess for view updates we'd like to have cl::ANY_BUT_PLEASE_ALWAYS_LEAVE_HINTS_TO_DEAD_ENDPOINTS_ANYWAY...

To sum up, this patch changes the behaviour if there are more than 1 destination node from `if any node is down, fail and write hints` to `if any node is up - it's fine, otherwise write hints`. Neither fits the case perfectly.

Reply:
When topology is changing CL is elevated and one response will not be enough to achieve it. For that matter calling timeout_cb() for MV is kind of OK. I never claimed this is a bug. But if we want to make cancellation mechanism generic we need to inject a failure here.